### PR TITLE
[ottf,tests] add ottf context switching and example concurrency test

### PR DESCRIPTION
**Pull request by timothytrippel**
This PR contains several commits that:

- increases the size of the FreeRTOS heap to get concurrency tests working again (the heap was too small causing a malloc error)
- enables (cooperative) context switching between FreeRTOS tasks through the use of the machine-mode ecall
- adds several FreeRTOS wrapper functions to the OTTF API that simplify the FreeRTOS interface by providing functions to:
    - create additional FreeRTOS tasks,
    - delete the currently running task,
    - cooperatively yield control flow between tasks, and
    - get the name of the task that is currently executing.
- adds an example test concurrency to showcase how to use the OTTF to launch, and context switch between several concurrent FreeRTOS tasks.

This partially addresses #8015

### CODE REVIEW

**Comment by timothytrippel** [comment]
CC @sha-ron @msfschaffner

File: sw/device/lib/testing/test_framework/ottf_main.h [hunk 1]
```diff
@@ -45,4 +52,58 @@ extern bool manufacturer_pre_test_hook(void);
...
+ * @param task_name A task identification string used to help debugging.
+ * @param task_stack_depth The amount of memory to reserve for the task's stack.
+ * @param task_priority The numerical priority of the task.
+ * @return A boolean encoding the success of the operation.
+ */
+inline bool ottf_task_create(TaskFunction_t task_function,
+                             const char *task_name,
+                             configSTACK_DEPTH_TYPE task_stack_depth,
+                             uint32_t task_priority) {
+  return xTaskCreate(task_function, task_name, task_stack_depth, NULL,
```
Comment (moidx) [inline sw/device/lib/testing/test_framework/ottf_main.h:77]: nit: from style guide, add `/*param_name=*/NULL` to all value parameters.

File: sw/device/lib/testing/test_framework/ottf_main.h [hunk 2]
```diff
@@ -45,4 +52,58 @@ extern bool manufacturer_pre_test_hook(void);
...
+ */
+inline void ottf_task_yield(void) { taskYIELD(); }
+
+/**
+ * Delete the calling FreeRTOS task.
+ *
+ * See the FreeRTOS `vTaskDelete` documentation for more details:
+ * https://www.freertos.org/a00126.html.
+ */
+inline void ottf_task_delete_self(void) { vTaskDelete(NULL); }
```
Comment (moidx) [inline sw/device/lib/testing/test_framework/ottf_main.h:99]: `/*param_name=*/NULL`

File: sw/device/lib/testing/test_framework/ottf_main.h [hunk 3]
```diff
@@ -45,4 +52,58 @@ extern bool manufacturer_pre_test_hook(void);
...
+
+/**
+ * Returns the name of the currently executing FreeRTOS task.
+ *
+ * See the FreeRTOS `pcTaskGetName` documentation for more details:
+ * https://www.freertos.org/a00021.html#pcTaskGetName.
+ */
+inline char *ottf_task_get_self_name(void) { return pcTaskGetName(NULL); }
```
Comment (moidx) [inline sw/device/lib/testing/test_framework/ottf_main.h:107]: `/*param_name=*/NULL`

**Review by moidx — APPROVED** [review]

File: sw/device/tests/example_concurrency_test.c [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+  // Place test code below.
+  // ***************************************************************************
+  LOG_INFO("Executing %s ...", ottf_task_get_self_name());
+  ottf_task_yield();
+  LOG_INFO("Continuing to execute %s ...", ottf_task_get_self_name());
+
+  // ***************************************************************************
+  // Delete the current task and never return.
+  // ***************************************************************************
+  ottf_task_delete_self();
```
[the remark this answers is not in the document]
Comment (timothytrippel) [reply sw/device/tests/example_concurrency_test.c]: good idea. done.

File: sw/device/tests/example_concurrency_test.c [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+ * according to the instructions below.
+ *
+ * This example demonstrates a concurrency test that spawns three FreeRTOS
+ * tasks, in addition to the `test_main` task. Each task is defined using a
+ * separate function that never returns, and deletes itself after executing its
+ * code. Since the priorities of each task are the same, yet higher than the
+ * priority of the "test_main" task, calling `ottf_task_yield()` will switch
+ * control flow between these tasks, until each task deletes itself. Then, the
+ * `test_main` task will continue executing, returning `true` when the overall
+ * test passes, triggering the OTTF to signal test execution has completed.
```
[the remark this answers is not in the document]
Comment (timothytrippel) [reply sw/device/tests/example_concurrency_test.c:18]: fixed.

File: sw/device/lib/testing/test_framework/ottf_main.h (hunk 1 above)
Comment (timothytrippel) [reply sw/device/lib/testing/test_framework/ottf_main.h]: fixed.

File: sw/device/lib/testing/test_framework/ottf_main.h (hunk 3 above)
Comment (timothytrippel) [reply sw/device/lib/testing/test_framework/ottf_main.h]: fixed.

File: sw/device/lib/testing/test_framework/FreeRTOSConfig.h [hunk 6]
```diff
@@ -26,7 +26,7 @@
 #define configCHECK_FOR_STACK_OVERFLOW 1
 #define configMINIMAL_STACK_SIZE 256  // in words
 #define configSTACK_DEPTH_TYPE uint16_t
-#define configTOTAL_HEAP_SIZE ((size_t)0x800u)
+#define configTOTAL_HEAP_SIZE ((size_t)0x8000u)
```
Comment (alphan) [inline sw/device/lib/testing/test_framework/FreeRTOSConfig.h:29]: nit: do we need to update any linker scripts?

File: sw/device/lib/testing/test_framework/ottf_macros.h [hunk 7]
```diff
@@ -9,5 +9,9 @@
 #define OTTF_NV_SCRATCH _non_volatile_scratch_start
 #define OTTF_HALF_WORD_SIZE (OTTF_WORD_SIZE / 2)
 #define OTTF_CONTEXT_SIZE (OTTF_WORD_SIZE * 30)
+#define OTTF_TASK_DELETE_SELF_OR_DIE \
+  ottf_task_delete_self();           \
+  abort();                           \
+  CHECK(false);
```
Comment (alphan) [inline sw/device/lib/testing/test_framework/ottf_macros.h:15]: Isn't `CHECK(false)` dead-code? I thought `abort()` is `while(true) { wfi }`

File: sw/device/lib/testing/test_framework/ottf_main.h [hunk 8]
```diff
@@ -7,8 +7,8 @@
 
 #include <stdbool.h>
 
-// This private header is included here so that OTTF users can include a single
-// header in their test application (the `ottf_main.h` header).
+#include "external/freertos/include/FreeRTOS.h"
+#include "external/freertos/include/task.h"
```
Comment (alphan) [inline sw/device/lib/testing/test_framework/ottf_main.h:11]: nit: including these headers makes the symbols in them available to clients, thus `kOttfFreeRtosMinStackSize` is more like an alias. Do you need the new functions to be inline, e.g. are the functions they call inline as well? If not, maybe they can be non-inline functions.

File: sw/device/lib/testing/test_framework/FreeRTOSConfig.h (hunk 6 above)
Comment (timothytrippel) [reply sw/device/lib/testing/test_framework/FreeRTOSConfig.h:29]: No we should not need to update any linker script as the FreeRTOS heap is an uninitialized statically allocated array that we mark with a specific [section label](https://cs.opensource.google/opentitan/opentitan/+/master:sw/device/lib/testing/test_framework/ottf_common.ld;drc=88f78ad969a50dc088008ac58f399f524288bc91;l=226). It gets placed in the bss section.

File: sw/device/lib/testing/test_framework/ottf_macros.h (hunk 7 above)
Comment (timothytrippel) [reply sw/device/lib/testing/test_framework/ottf_macros.h:15]: yes, fixed in #15249.

File: sw/device/lib/testing/test_framework/ottf_main.h (hunk 8 above)
Comment (timothytrippel) [reply sw/device/lib/testing/test_framework/ottf_main.h:11]: The original intent of the inline nature of these wrapper functions was to reduce the overhead of these additional function calls in tests. As we discussed offline, moving these to non-inline functions is probably better for isolation of FreeRTOS code. Fixed in #15249.

**Pull request merged**